modules/rostests/apitests/atl/CAtlList.cpp

```diff
@@ -113,9 +113,54 @@ test_SwapElements()
     ok_list(list, "3,1,2,");
 }
 
+static void
+test_AppendListToTail()
+{
+    CAtlList<int> list;
+    list.AddTail(1);
+    list.AddTail(2);
+    list.AddTail(0);
+    ok_list(list, "1,2,0,");
+    
+    CAtlList<int> list_tail;
+    list_tail.AddTail(8);
+    list_tail.AddTail(1);
+    list_tail.AddTail(0);
+    ok_list(list_tail, "8,1,0,");
+    
+    list.AddTailList(&list_tail);
+    ok_list(list, "1,2,0,8,1,0,");
+    
+    list_tail.AddTailList(&list);
+    ok_list(list_tail, "8,1,0,1,2,0,8,1,0,");
+}
+
+static void
+test_AppendListToHead()
+{
+    CAtlList<int> list_head;
+    list_head.AddHead(0);
+    list_head.AddHead(0);
+    list_head.AddHead(2);
+    ok_list(list_head, "2,0,0,");
+    
+    CAtlList<int> list;
+    list.AddHead(8);
+    list.AddHead(9);
+    list.AddHead(7);
+    ok_list(list, "7,9,8,");
+    
+    list.AddHeadList(&list_head);
+    ok_list(list, "2,0,0,7,9,8,");
+    
+    list_head.AddHeadList(&list);
```

Comment:
Can the ATL list add itself? For example,

`list_head.AddHeadList(&list_head);`

Reply:
I have tried that and it caused the test to crash without reporting any failures. But that works with official ATL from Microsoft.

Reply:
Okay, Let's save and remember the ending position.

Reply:
> Okay, Let's save and remember the ending position.

And here's why: we have a memory leak!
![image](https://github.com/user-attachments/assets/53554045-5c9e-40f3-aec3-372039289618)

Reply:
The list is adding itself at the tail position.
That is an infinite loop.

Reply:
At least in the ATL version with VS2010, they don't appear to do anything special to avoid self-linking.

Reply:
Not fully clear from the previous discussion:
How does your self-linking test behave when using MS ATL @SigmaTel71 ?

Reply:
> How does your self-linking test behave when using MS ATL @SigmaTel71 ?

If you do `list_head.AddHeadList(&list_head)` to the list with *two* or more items, MS ATL will fall into the infinite loop and drain RAM.

Reply:
We don't have an [ATLENSURE](https://learn.microsoft.com/en-us/cpp/atl/reference/debugging-and-error-reporting-macros?view=msvc-170#atlensure) macro to assert in release builds as well, so if there are any alternatives, I would like to know them.